Sources/Soto/Extensions/S3/S3+presignedPost.swift

```diff
@@ -0,0 +1,251 @@
+        let shortDate = self.shortDateFormat(date: date)
+
+        let clientCredentials = try await client.getCredential()
+        let presignedPostCredential = try await getPresignedPostCredential(date: shortDate, accessKeyId: clientCredentials.accessKeyId)
+
+        var keyCondition: PostPolicyCondition
+        let suffix = "${filename}"
+        if key.hasSuffix(suffix) {
+            keyCondition = .rule("starts-with", "$key", String(key.dropLast(suffix.count)))
+        } else {
+            keyCondition = .match("key", key)
+        }
+
+        // Add required conditions
+        conditions.append(.match("bucket", bucket))
+        conditions.append(keyCondition)
+        conditions.append(.match("x-amz-algorithm", algorithm))
+        conditions.append(.match("x-amz-date", longDate))
+        conditions.append(.match("x-amz-credential", presignedPostCredential))
+
+        // Add required fields
+        fields["key"] = key
+        fields["x-amz-algorithm"] = algorithm
+        fields["x-amz-date"] = longDate
+        fields["x-amz-credential"] = presignedPostCredential
+
+        // Create the policy and add to fields
+        let policy = PostPolicy(expiration: date.addingTimeInterval(expiresIn), conditions: conditions)
+        let stringToSign = try policy.stringToSign()
+
+        fields["Policy"] = stringToSign
+
+        // Create the signature and add to fields
+        let signingKey = try await signingKey(date: shortDate, secretAccessKey: clientCredentials.secretAccessKey)
+        let signature = try await getSignature(policy: stringToSign, signingKey: signingKey)
+        fields["x-amz-signature"] = signature
+
+        // Create the response
+        let presignedPostResponse = PresignedPostResponse(url: url, fields: fields)
+
+        return presignedPostResponse
+    }
+
+    func signingKey(date: String, secretAccessKey: String) async throws -> SymmetricKey {
+        let credentials = try await client.getCredential()
+        let name = config.signingName
+        let region = config.region.rawValue
+
+        let kDate = HMAC<SHA256>.authenticationCode(for: [UInt8](date.utf8), using: SymmetricKey(data: Array("AWS4\(credentials.secretAccessKey)".utf8)))
+        let kRegion = HMAC<SHA256>.authenticationCode(for: [UInt8](region.utf8), using: SymmetricKey(data: kDate))
+        let kService = HMAC<SHA256>.authenticationCode(for: [UInt8](name.utf8), using: SymmetricKey(data: kRegion))
+        let kSigning = HMAC<SHA256>.authenticationCode(for: [UInt8]("aws4_request".utf8), using: SymmetricKey(data: kService))
+        return SymmetricKey(data: kSigning)
+    }
+
+    func getSignature(policy: String, signingKey key: SymmetricKey) async throws -> String {
+        let signature = HMAC<SHA256>.authenticationCode(for: [UInt8](policy.utf8), using: key).hexDigest()
+        return signature
+    }
+
+    func getPresignedPostCredential(date: String, accessKeyId: String) async throws -> String {
```

Comment:
You'll need to fix the call sites as well
```suggestion
    func getPresignedPostCredential(date: String, accessKeyId: String) -> String {
```